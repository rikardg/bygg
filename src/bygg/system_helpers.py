"""
System-near helper functions. Only import things from the standard library here; this
way these helpers can be used also for utility scripts.
"""

import contextlib
import errno
import os
import pty
import shlex
import signal
import subprocess
from typing import Generator


class ProcessGenerator:
    """
    Class that wraps a process return value and a generator that yields the output of a
    process line by line.

    Attributes
    ----------
    returncode : int
        Process return code.
    generator : Generator
        Generator that yields the output of the process line by line.

    Yields
    ------
    str
        Output of the process.
    """

    returncode: int
    """Process return code."""

    generator: Generator
    """Generator that yields the output of the process line by line."""

    def __init__(self):
        self.returncode = -1

    def __iter__(self):
        self.value = yield from self.generator


def subprocess_tty(cmd, encoding="utf-8", timeout=10, **kwargs):
    """
    Wrapper around subprocess.Popen that sets up the process as if it were running in a
    TTY and returns a generator that yields the stdout of the process line by line.

    Parameters
    ----------
    cmd :
        Passed directly to subprocess.Popen; see its documentation for details.
    encoding : str, optional
        Encoding to use when reading stdout, by default "utf-8".
    timeout : int, optional
        Timeout to use when polling the child process to see if it has exited, by
        default 10 seconds.

    Returns
    -------
    ProcessGenerator
        Generator that yields the stdout of the child process line by line. After the
        generator is exhausted, the returncode attribute of the generator will be set to
        the exit code of the child process.

    Yields
    ------
    str
        Lines of stdout from the child process.

    Raises
    ------
    subprocess.TimeoutExpired
        Re-raised if the child process did not exit within the set timeout after having
        received both terminate and kill signals.

    Can also raise other exceptions from subprocess.Popen.
    """

    generator = ProcessGenerator()

    def subprocess_iterator():
        # From https://stackoverflow.com/a/77387332
        m, s = pty.openpty()
        p = subprocess.Popen(cmd, stdout=s, stderr=s, **kwargs)
        os.close(s)

        try:
            for line in open(m, encoding=encoding):
                if not line:  # EOF
                    break
                yield line
        except OSError as e:
            if e.errno != errno.EIO:  # EIO also means EOF
                raise
        finally:
            if p.poll() is None:
                p.send_signal(signal.SIGINT)
                try:
                    p.wait(timeout)
                except subprocess.TimeoutExpired:
                    p.terminate()
                    try:
                        p.wait(timeout)
                    except subprocess.TimeoutExpired:
                        p.kill()
                        raise
            p.wait()
            generator.returncode = p.returncode

    generator.generator = subprocess_iterator()
    return generator


def subprocess_tty_print(cmd, encoding="utf-8", timeout=10, **kwargs):
    """
    Thin wrapper around subprocess_tty that prints its output line by line. See
    subprocess_tty for details.


    Parameters
    ----------
    See subprocess_tty.

    Returns
    -------
    int
        The status code from the subprocess.
    """
    proc = subprocess_tty(cmd, encoding, timeout, **kwargs)
    for line in proc:
        print(line.rstrip())
    return proc.returncode


class ExitCode(int):
    """
    An int with a customised __bool__ method that considers 0 to be truthy. This is
    useful for working with subprocess return codes.
    """

    def __bool__(self):
        return self == 0


def call(cmd, encoding="utf-8", timeout=10, **kwargs) -> ExitCode:
    """
    Convenience wrapper around subprocess_tty_print; the command to be executed can be
    given as string that will be split instead of as an array. Output will be printed
    line by line.

    Parameters
    ----------
    cmd : str
        The command to be executed. Will be run through shlex.split().

    See subprocess_tty_print and subprocess_tty for the other parameters.

    Returns
    -------
    ExitCode
        The exit code from the subprocess. ExitCode is a subclass of int that can be
        used as a boolean; subprocess success (0) is considered truthy.

    Raises
    ------
    See subprocess_tty_print and subprocess_tty.
    """

    return ExitCode(subprocess_tty_print(shlex.split(cmd), encoding, timeout, **kwargs))


@contextlib.contextmanager
def change_dir(dir: str | os.PathLike | None):
    old_dir = os.getcwd()
    try:
        if dir is not None:
            os.chdir(dir)
        yield
    finally:
        os.chdir(old_dir)
