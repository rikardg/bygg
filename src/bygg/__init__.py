from .core.action import Action, ActionContext, WorkChannel, action
from .core.common_types import CommandStatus

__all__ = [
    "Action",
    "ActionContext",
    "CommandStatus",
    "WorkChannel",
    "action",
]
