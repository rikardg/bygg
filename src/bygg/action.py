from dataclasses import dataclass
from typing import Callable, Iterable, Literal, Optional, Set, Tuple

from bygg.common_types import CommandStatus

SchedulingType = Literal["in-process", "processpool"]

# Function that returns a string that is included in the dependency digest. Returning
# None causes the value to be ignored.
DynamicDependency = Callable[[], str | None]


@dataclass
class ActionContext:
    name: str
    message: str | None
    inputs: Set[str]
    outputs: Set[str]
    dependencies: Set[str]
    dynamic_dependency: Optional[DynamicDependency]
    is_entrypoint: bool
    scheduling_type: SchedulingType


Command = Callable[
    [ActionContext],
    CommandStatus,
]


class Action(ActionContext):
    """
    An action in the build graph.

    name: The name of the action.
    message (optional): A message to display when the action is run.
    inputs (optional): A list of paths that are inputs to the action.
    outputs (optional): A list of paths that are outputs of the action.
    dependencies (optional): A list of action names that this action depends on.
    is_entrypoint (optional): Whether this action is an entrypoint to the build graph.
    command (optional): Instance of one of the Command types to run.

    """

    command: Command | None
    dependency_files: Set[str]

    def __init__(
        self,
        name: str,
        description: str | None = None,
        message: str | None = None,
        inputs: Optional[Iterable[str]] = None,
        outputs: Optional[Iterable[str]] = None,
        dependencies: Optional[Iterable[str]] = None,
        dynamic_dependency: Optional[DynamicDependency] = None,
        is_entrypoint: bool = False,
        command: Command | None = None,
        scheduling_type: SchedulingType = "processpool",
    ):
        from bygg.scheduler import scheduler

        self.name = name
        self.description = description
        self.message = message
        self.inputs = {*inputs} if inputs else set()
        self.outputs = {*outputs} if outputs else set()
        self.dependencies = {*dependencies} if dependencies else set()
        self.dynamic_dependency = dynamic_dependency
        self.is_entrypoint = is_entrypoint
        self.command = command
        self.scheduling_type = scheduling_type

        self.dependency_files = set()

        scheduler.build_actions[name] = self

    def __repr__(self):
        return f"""Action(name={self.name}, inputs=({
                ', '.join([str(i) for i in self.inputs])}), outputs=({
                ', '.join([str(o) for o in self.outputs])}), dependencies=({
                ', '.join(self.dependencies)}) is_entrypoint={
            self.is_entrypoint})"""

    def __str__(self):
        return self.__repr__()


def action(
    name: str,
    *,
    message: Optional[str] = None,
    inputs: Optional[Iterable[str]] = None,
    outputs: Optional[Iterable[str]] = None,
    dependencies: Optional[Iterable[str]] = None,
    dynamic_dependency: Optional[DynamicDependency] = None,
    is_entrypoint: bool = False,
):
    """Decorator for creating an Action from a function."""

    def create_action(func: Callable):
        Action(
            name,
            message,
            inputs=inputs,
            outputs=outputs,
            dependencies=dependencies,
            dynamic_dependency=dynamic_dependency,
            is_entrypoint=is_entrypoint,
            command=func,
        )

    return create_action


def action_set(
    base_name: str,
    *,
    message: Optional[str] = None,
    file_pairs: Iterable[Tuple[str, str]],
    dependencies: Optional[Iterable[str]] = None,
    is_entrypoint: bool = False,
):
    """
    Decorator for creating individual Actions from a list of input-output file pairs and
    a function. Creates one Action that depends on all of the individual Actions. It has
    name = base_name, which in turn can be used as a dependency for other Actions.
    """

    def create_actions(func: Callable):
        action_list = []
        for input_file, output_file in file_pairs:
            action_name = f"{base_name}_{output_file}"
            action_list.append(action_name)
            Action(
                action_name,
                message,
                inputs=[input_file],
                outputs=[output_file],
                dependencies=dependencies,
                is_entrypoint=is_entrypoint,
                command=func,
            )
        Action(base_name, f"Action set {base_name}", dependencies=action_list)

    return create_actions
