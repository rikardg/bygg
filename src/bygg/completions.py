import argparse
from pathlib import Path
import shutil
import sys
import textwrap
from typing import Any, Set

from argcomplete.completers import DirectoriesCompleter
from argcomplete.finders import CompletionFinder

from bygg.output import output_plain


class ByggfileDirectoriesCompleter(DirectoriesCompleter):
    """
    A completer for directories that contain Bygg files.
    """

    def __call__(self, prefix, **kwargs):
        directories = super().__call__(prefix, **kwargs)
        byggfile_dirs = set()
        for dir in directories:
            byggfile_dirs |= {
                str(b.parent)
                for b in Path(dir).rglob("Byggfile.*")
                if b.suffix in {".py", ".yml"}
            }
        return sorted(list(byggfile_dirs))


def construct_already_there(
    parser: argparse.ArgumentParser | Any, options: Set[str]
) -> Set[str]:
    """
    Construct a set of options that are already present among the command line
    arguments. Adds both the short and long versions if they exist.
    """
    already_there_set: Set[str] = set()
    for action in parser._get_optional_actions():
        option_strings = set(action.option_strings)
        if option_strings & options:
            already_there_set.update(option_strings)
    return already_there_set


singular_options = {
    "--completions",
    "--dump-schema",
    "--help",
    "--version",
    "-h",
    "-v",
}

directory_option = {"-C", "--directory"}


class ByggCompletionFinder(CompletionFinder):
    """
    Override _get_completions because we want custom completion patterns that are not
    expressed in argparse.
    """

    def _get_completions(
        self, comp_words, cword_prefix, cword_prequote, last_wordbreak_pos
    ):
        comp_words_set = set(comp_words)

        # Stop completions if we already have a "singular" option, i.e. an option that
        # should not be followed by anything else.
        if comp_words_set & singular_options:
            return []

        # Get completions from argparse.
        completions = set(
            super()._get_completions(
                comp_words, cword_prefix, cword_prequote, last_wordbreak_pos
            )
        )

        # If the last word is a directory, the default completions will be from the
        # directories completer, so return these.
        if comp_words[-1] in directory_option:
            return completions

        action_completions = list(filter(lambda x: not x.startswith("-"), completions))
        already_there = construct_already_there(self._parser, comp_words_set)

        if comp_words_set & {"--tree", "--clean"}:
            # Actions if we have them.
            if action_completions and not cword_prefix.startswith("-"):
                return action_completions
            return completions & directory_option - already_there

        if comp_words_set & {"--list"}:
            # Only complete directories.
            return completions & directory_option - already_there

        # Return action completions if they exist and user has not entered a '-'.
        if action_completions and not cword_prefix.startswith("-"):
            return action_completions

        return completions - already_there


def do_completion(parser: argparse.ArgumentParser):
    completer = ByggCompletionFinder()
    completer(parser)


def generate_shell_completions(print_and_exit: bool):
    bygg_bin = Path(sys.argv[0])
    bygg_completions = bygg_bin.resolve().parent / "_bygg_completions.sh"
    if not bygg_completions.exists():
        import argcomplete.shell_integration

        with open(bygg_completions, "w") as f:
            f.write(argcomplete.shell_integration.shellcode([str(bygg_bin.name)]))

    if print_and_exit:
        terminal_cols, _ = shutil.get_terminal_size()
        indent_string = "# "
        width = min(terminal_cols, 80) - len(indent_string)
        description = f"""\
            A shell completions script has been generated in {bygg_completions}. It will
            pick up the bygg that is in PATH. Add the following line to your shell's
            startup script to load completions:
        """
        output_plain(
            textwrap.indent(
                textwrap.fill(
                    textwrap.dedent(description),
                    width=width,
                    break_long_words=False,
                    break_on_hyphens=False,
                ),
                indent_string,
            )
        )
        output_plain(f"\nsource {bygg_completions}")

        sys.exit(0)
