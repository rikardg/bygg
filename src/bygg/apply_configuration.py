from pathlib import Path
import shutil
import subprocess
import sys
import time
from typing import List

import rich
import rich.status

from bygg.action import Action
from bygg.configuration import (
    PYTHON_INPUTFILE,
    ByggFile,
    Environment,
    load_python_build_file,
)
from bygg.digest import calculate_string_digest
from bygg.output import output_error, output_info, output_plain
from bygg.scheduler import scheduler
from bygg.util import create_shell_command


def calculate_environment_hash(environment: Environment) -> str:
    requirements: List[str] = []
    requirements.append(environment.shell)
    if environment.inputs:
        for input in environment.inputs:
            with open(input, "r") as f:
                requirements += f.readlines()
    return calculate_string_digest(" ".join(requirements))


loading_python_build_file = rich.status.Status(
    "[cyan]Executing Python build file", spinner="dots"
)


def setup_environment(environment: Environment):
    venv_path = Path(
        environment.venv_directory if environment.venv_directory else ".venv"
    )

    environment_hash = calculate_environment_hash(environment)
    environment_hash_file = venv_path / "bygg_environment_hash.txt"
    if environment_hash_file.exists():
        with open(environment_hash_file, "r") as f:
            if f.read() == environment_hash:
                return True
        environment_hash_file.unlink()

    # TODO should we remove the whole venv or just trust pip to do the right thing? For
    # now, remove the venv.

    if venv_path.exists():
        output_info(f"Replacing venv at {venv_path}")
        shutil.rmtree(venv_path)

    output_info("Setting up environment")
    process = subprocess.run(
        environment.shell,
        shell=True,
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        encoding="utf-8",
    )

    output_plain(process.stdout)

    if process.returncode != 0:
        output_error("Error while creating virtual environment:")
        output_plain(process.stdout)
        sys.exit(1)

    with open(environment_hash_file, "w") as f:
        f.write(environment_hash)

    return True


def should_restart_with(environment: Environment) -> str | None:
    venv_bin_path = Path(environment.venv_directory) / "bin"

    # # TODO: we could check here if the new venv is using the same Python binary that
    # # we're already running under, and if so just activate the venv.

    is_relative = Path(sys.executable).is_relative_to(venv_bin_path.resolve())
    if not is_relative:
        return str((venv_bin_path / "bygg"))
    return None


def apply_configuration(
    configuration: ByggFile | None,
    environment_name: str | None,
    is_restarted_with_env: str | None,
) -> str | None:
    """Returns the path of the bygg install to restart with if a restart is needed."""
    if configuration:
        if not is_restarted_with_env and environment_name:
            if environment_name in configuration.environments:
                environment = configuration.environments[environment_name]
                setup_environment(environment)

                restart_with = should_restart_with(environment)
                if restart_with is not None:
                    return restart_with

        for action in configuration.actions:
            # TODO Hack until we can iterate over the dependency graph in scheduler.prepare_run:
            if is_restarted_with_env and action.environment != is_restarted_with_env:
                continue

            shell_command = (
                create_shell_command(action.shell, action.message)
                if action.shell
                else None
            )
            Action(
                action.name,
                is_entrypoint=bool(action.is_entrypoint),
                inputs=action.inputs,
                outputs=action.outputs,
                dependencies=action.dependencies,
                command=shell_command,
            )

    # Evaluate the Python build file:

    python_build_file = PYTHON_INPUTFILE
    if configuration and environment_name:
        environment = configuration.environments.get(environment_name, None)
        if environment:
            python_build_file = environment.byggfile

    t0 = time.time()
    action_count = len(scheduler.build_actions)

    with loading_python_build_file:
        load_python_build_file(python_build_file)
    output_info(
        f"{len(scheduler.build_actions) - action_count} actions registered in "
        f"{time.time() - t0:.2f} seconds."
    )

    return None
