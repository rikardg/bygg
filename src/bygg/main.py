# PYTHON_ARGCOMPLETE_OK

import argparse
from dataclasses import dataclass
import os
import shutil
import stat
import subprocess
import sys
import time
from typing import Any, List

import argcomplete

from bygg.action import Action
from bygg.apply_configuration import apply_configuration
from bygg.configuration import (
    PYTHON_INPUTFILE,
    YAML_INPUTFILE,
    ByggFile,
    dump_schema,
    read_config_file,
)
from bygg.output import output_error, output_info, output_ok, output_warning
from bygg.runner import runner
from bygg.scheduler import scheduler
from bygg.status_display import (
    failed_checks,
    on_job_status,
    on_runner_status,
    output_check_results,
    progress,
)


def get_job_count_limit():
    try:
        # Use os.sched_getaffinity where available (on U**X):
        # https://stackoverflow.com/a/55423170
        return len(os.sched_getaffinity(0))
    except AttributeError:
        count = os.cpu_count()
        assert count is not None
        return count


def init_status_listeners():
    """
    Set up status listeners.
    """
    runner.job_status_listener = on_job_status
    runner.runner_status_listener = on_runner_status


def build(
    actions: List[str],
    job_count: int | None,
    always_make: bool,
    check: bool,
) -> bool:
    """
    actions: The actions to build.

    job_count: The number of jobs to run simultaneously. None means to use the number of
    available cores.

    always_make: If True, all actions will be built, even if they are up to date.

    check: If True, apply various checks:

    * Check that the inputs and outputs of all actions will be checked for consistency.
      A job that runs later must not have files as output that are inputs to a job that
      runs earlier.
    """
    try:
        init_status_listeners()

        max_workers = get_job_count_limit() if job_count is None else job_count

        for action in actions:
            t1 = time.time()
            output_info(f"Building action '{action}':")

            progress.start()
            scheduler.start_run(
                action,
                always_make=always_make,
                check=check,
            )
            status = runner.start(max_workers)
            progress.disable
            progress.stop()
            scheduler.shutdown()

            if status:
                output_ok(f"Action '{action}' completed in {time.time() - t1:.2f} s.")
            else:
                output_error(
                    f"Action '{action}' failed after {time.time() - t1:.2f} s."
                )

            # cs = build(action)
            # if cs is None:
            #     rich.print(f"Action '{action}' is up to date.")
            # else:
            #     rich.print(f"Action '{action}' completed with return code {cs.rc}.")
            #     rich.print(
            #         f"{len(cs.changed_files)} files changed in {time() - t1:.2f} s."
            #     )
            # rich.print("=========================================")
    except KeyboardInterrupt:
        output_warning("\n[yellow]Build was interrupted by user.[/yellow]")
        return False
    except KeyError as e:
        output_error(f"Error: Action '{e}' not found.")
        return False
    finally:
        progress.stop()
        scheduler.shutdown()

    if check and failed_checks:
        return output_check_results()

    return True


def clean(actions: List[str]):
    try:
        init_status_listeners()

        for action in actions:
            output_info(f"Cleaning action '{action}':")
            scheduler.prepare_run(action)
            for job_name in scheduler.job_graph.get_all_jobs():
                job = scheduler.build_actions.get(job_name, None)
                if job is None:
                    continue
                for output in job.outputs:
                    try:
                        s = os.stat(output)
                        if stat.S_ISREG(s.st_mode):
                            os.unlink(output)
                        elif stat.S_ISDIR(s.st_mode):
                            output_info(f"Removing directory: {output}")
                            shutil.rmtree(output)
                    except FileNotFoundError:
                        pass
            scheduler.shutdown()
    except KeyboardInterrupt:
        output_warning("Build was interrupted by user.")
        return False
    except KeyError as e:
        output_error(f"Error: Action '{e}' not found.")
        return False
    finally:
        progress.stop()
        scheduler.shutdown()

    return True


def get_entrypoints(configuration: ByggFile | None) -> List[Action]:
    entrypoints = []
    if configuration:
        entrypoints += [x for x in configuration.actions if x.is_entrypoint]

    if not configuration or not configuration.environments:
        entrypoints += [x for x in scheduler.build_actions.values() if x.is_entrypoint]

    return entrypoints


def list_actions(configuration: ByggFile | None) -> bool:
    entrypoints = get_entrypoints(configuration)

    if entrypoints:
        max_width = max([len(x.name) for x in entrypoints])
        output_info("Available actions:")
        for action in sorted(entrypoints, key=lambda x: x.name):
            output_info(
                f"  {action.name: <{max_width}} : {action.description if action.description else 'No description'}"
            )
        return True
    else:
        program_name = os.path.basename(sys.argv[0])
        output_error("Loaded build files but no entrypoints were found.")
        output_error(f"Type `{program_name} --help` for help.")
        return False


def print_no_actions_text(configuration: ByggFile | None):
    return list_actions(configuration)


def print_version():
    import importlib.metadata

    output_info(f"{__package__} {importlib.metadata.version(__package__)}")


MAKE_COMPATIBLE_PANEL = "(Roughly) Make-compatible options"


def list_actions_and_exit(configuration: ByggFile | None):
    status = list_actions(configuration)
    sys.exit(0) if status else sys.exit(1)


def dispatcher(args: argparse.Namespace):
    """
    A build tool written in Python, where all actions can be written in Python.
    """
    if args.dump_schema:
        dump_schema()
        sys.exit(0)

    if args.version:
        print_version()
        sys.exit(0)

    directory = args.directory[0] if args.directory else None
    is_restarted_with_env = (
        args.is_restarted_with_env[0] if args.is_restarted_with_env else None
    )

    if directory and not is_restarted_with_env:
        directory_arg = args.directory[0]
        output_info(f"Entering directory '{directory_arg}'")
        os.chdir(directory_arg)

    if not os.path.isfile(PYTHON_INPUTFILE) and not os.path.isfile(YAML_INPUTFILE):
        output_error("No build files found.")
        sys.exit(1)

    configuration = read_config_file()

    if args.list:
        list_actions_and_exit(configuration)

    action_partitions = partition_actions(configuration, args.actions)

    if not action_partitions:
        if os.path.isfile(PYTHON_INPUTFILE):
            # No configuration file, so load the Python build file directly.
            apply_configuration(None, None, None)
            # TODO will not be needed once we allow configuring default action also in
            # Python.
            if not args.actions:
                list_actions_and_exit(configuration)
            status = do_dispatch(args, args.actions)
        else:
            status = print_no_actions_text(configuration)
        if status:
            sys.exit(0)
        sys.exit(1)

    for partition in action_partitions:
        env = partition.environment_name
        # Check if we should restart with another Python interpreter (e.g. from a
        # virtualenv):
        restart_with = apply_configuration(configuration, env, is_restarted_with_env)

        if restart_with is not None and not is_restarted_with_env:
            exec_list = construct_exec_list(args, restart_with, partition)
            try:
                process = subprocess.run(exec_list, encoding="utf-8")
                if process.returncode != 0:
                    sys.exit(process.returncode)
            except FileNotFoundError:
                output_error(f"Error: Could not restart with '{restart_with}'.")
                output_warning(
                    "Please make sure that bygg is in your pip requirements list for this environment."
                )
                sys.exit(1)
        else:
            status = do_dispatch(args, partition.actions)
            if not status:
                sys.exit(1)


def do_dispatch(args: argparse.Namespace, actions: List[str]) -> bool:
    # Analysis implies building:
    always_make = args.always_make or args.check
    if args.clean:
        status = clean(actions)
    else:
        jobs = int(args.jobs) if args.jobs else None
        status = build(actions, jobs, always_make, args.check)

    return status


@dataclass
class ActionPartition:
    """
    environment_name: The name of the environment that the actions should be run in.
    None means the implicit default environment, i.e. typically the base process.

    actions: The actions that should be run in the environment.
    """

    environment_name: str | None
    actions: List[str]


def construct_exec_list(
    args: argparse.Namespace, restart_with: str, partition: ActionPartition
):
    exec_list = [restart_with, *partition.actions]
    for k, v in vars(args).items():
        if k == "actions":
            continue
        elif v is True:
            exec_list.append(f"--{k}")
        elif v:
            exec_list.append(f"--{k} {v}")
    if partition.environment_name:
        exec_list.append("--is_restarted_with_env")
        exec_list.append(partition.environment_name)
    return exec_list


def partition_actions(
    configuration: ByggFile | None,
    actions: List[str] | None,
) -> List[ActionPartition] | None:
    """
    Partition the actions into groups that should be run in the same environment.
    """

    # TODO: Implement this.

    resolved_actions = actions if actions else []

    if configuration:
        if not actions and configuration.settings.default_action is not None:
            # Resolve to default action:
            resolved_actions += [configuration.settings.default_action]
            return [ActionPartition("default", resolved_actions)]

        # Put consecutive actions with the same environment in the same partition:
        action_partitions = []
        action_dict = {a.name: a for a in configuration.actions}
        current_partition: ActionPartition | None = None
        for action in [action_dict[a] for a in resolved_actions]:
            if (
                current_partition is None
                or current_partition.environment_name != action.environment
            ):
                current_partition = ActionPartition(action.environment, [action.name])
                action_partitions.append(current_partition)
            else:
                current_partition.actions.append(action.name)

        return action_partitions


def entrypoint_completions(prefix, parsed_args, **kwargs):
    entrypoints = get_entrypoints(read_config_file())
    return {x.name: x.message for x in entrypoints}


def create_argument_parser():
    parser = argparse.ArgumentParser(
        formatter_class=argparse.RawDescriptionHelpFormatter,
        description="""
A build tool written in Python, where all actions can be written in Python.

Build the default action:
 %(prog)s

Build ACTION:
 %(prog)s ACTION

Clean ACTION:
 %(prog)s ACTION --clean

List available actions:
 %(prog)s --list
""",
    )

    # Use Any to get around type checking for argcomplete:
    actions_action: Any = parser.add_argument(
        "actions",
        nargs="*",
        default=None,
        help="Entrypoint actions to operate on.",
    )

    actions_action.completer = entrypoint_completions

    # Only add completions for the actions:
    argcomplete.autocomplete(parser, exclude=["-h", "--help"])

    parser.add_argument(
        "-v", "--version", action="store_true", help="Show version string and exit."
    )

    parser.add_argument(
        "--is_restarted_with_env",
        nargs=1,
        type=str,
        default=None,
        help=argparse.SUPPRESS,
    )
    # Commands that operate on the build setup:
    build_setup_wrapper_group = parser.add_argument_group(
        "Commands that operate on the build setup"
    )  # add_mutually_exclusive_group doesn't accept a title, so wrap it in a regular group.
    build_setup_group = build_setup_wrapper_group.add_mutually_exclusive_group()
    build_setup_group.add_argument(
        "--clean",
        action="store_true",
        help="Clean the outputs of the specified actions.",
    )
    build_setup_group.add_argument(
        "-l",
        "--list",
        action="store_true",
        help="List available actions.",
    )
    # Some arguments inspired by Make:
    make_group = parser.add_argument_group("Make-like arguments")
    make_group.add_argument(
        "-C",
        "--directory",
        nargs=1,
        type=str,
        default=None,
        help="Change to the specified directory.",
    )
    make_group.add_argument(
        "-j",
        "--jobs",
        nargs="?",
        type=int,
        default=None,
        help="Specify the number of jobs to run simultaneously. None means to use the number of available cores.",
    )
    make_group.add_argument(
        "-B",
        "--always-make",
        action="store_true",
        help="Always build all actions.",
    )

    # Analyse and verify:
    analyse_group = parser.add_argument_group(
        "Analyse and verify",
        "Arguments in this group will add more analysis to the build process. Actions will be built and the analysis result will be reported.",
    )
    analyse_group.add_argument(
        "--check",
        action="store_true",
        help="Perform various checks on the action tree. Implies -B",
    )

    # Meta arguments:
    meta_group = parser.add_argument_group("Meta arguments")
    meta_group.add_argument(
        "--dump-schema",
        action="store_true",
        help="Generate a JSON Schema for the Byggfile.yml files. The schema will be printed to stdout.",
    )

    return parser


def main():
    parser = create_argument_parser()
    args = parser.parse_args()

    try:
        return dispatcher(args)
    except KeyboardInterrupt:
        output_warning("Interrupted by user. Aborting.")
        return 1


if __name__ == "__main__":
    main()
