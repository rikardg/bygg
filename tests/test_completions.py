from pathlib import Path

from bygg.completions import ByggCompletionFinder
from bygg.main import create_argument_parser
from bygg.system_helpers import change_dir
import pytest


def dummy_exit_method(status):
    pass


def open_raise(*args, **kwargs):
    raise NotImplementedError


@pytest.fixture
def completion_tester(monkeypatch, tmp_path):
    def tester(testcase):
        prefix = "bygg"
        prefixed_testcase = f"{prefix} {testcase}"
        outfile = tmp_path / "completion.out"

        # Simulate being run from the completion script:
        monkeypatch.setenv("_ARGCOMPLETE", "1")
        monkeypatch.setenv("COMP_LINE", prefixed_testcase)
        monkeypatch.setenv("COMP_POINT", str(len(prefixed_testcase)))
        monkeypatch.setenv("_ARGCOMPLETE_STDOUT_FILENAME", str(outfile))
        # Patch os.fdopen to raise NotImplementedError; this causes argcomplete to use
        # stderr for debug output instead of opening file descriptor 9, which causes pytest
        # to get its knickers in a twist.
        monkeypatch.setattr("os.fdopen", open_raise)

        parser = create_argument_parser()
        completer = ByggCompletionFinder()

        with open(outfile, "w", encoding="utf-8"):
            completer(parser, exit_method=dummy_exit_method)

        with open(outfile, "r", encoding="utf-8") as f:
            return f.read()

    return tester


args = [
    # Complete simple arguments:
    ("--ver", "--version "),
    ("--cl", "--clean "),
    ("--h", "--help "),
    # No completions after these arguments:
    ("-h ", ""),
    ("--help ", ""),
    ("-v ", ""),
    ("--version ", ""),
    ("--dump-schema ", ""),
    ("--completions ", ""),
]


@pytest.mark.parametrize("arg", args, ids=lambda x: x[0])
def test_completions(completion_tester, arg):
    testcase, correct_result = arg
    testresult = completion_tester(testcase)
    assert correct_result == testresult


# Directories:
directories_completions = [
    "-C ",
    "--directory ",
]


@pytest.mark.parametrize("arg", directories_completions, ids=lambda x: x)
def test_directory_completions(completion_tester, arg, snapshot):
    testcase = arg
    testresult = completion_tester(testcase)
    assert sorted(testresult.split("\x0b")) == snapshot


actions_completions = [p for p in Path("examples").iterdir() if p.is_dir()]


@pytest.mark.parametrize("arg", actions_completions, ids=lambda x: str(x.name))
def test_actions_completions(completion_tester, arg, snapshot, clean_bygg_tree):
    with change_dir(clean_bygg_tree):
        dir = str(arg)
        testcase = f"-C {dir} "
        testresult = completion_tester(testcase)
        assert sorted(testresult.split("\x0b")) == snapshot

        testcase = f"--directory {dir} "
        testresult = completion_tester(testcase)
        assert sorted(testresult.split("\x0b")) == snapshot

        with change_dir(dir):
            testresult = completion_tester("")
            assert sorted(testresult.split("\x0b")) == snapshot
