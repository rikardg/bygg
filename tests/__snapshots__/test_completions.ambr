# serializer version: 1
# name: test_actions_completions[checks]
  list([
    'all_checks',
    'check_inputs_outputs',
    'output_file_missing',
  ])
# ---
# name: test_actions_completions[checks].1
  list([
    'all_checks',
    'check_inputs_outputs',
    'output_file_missing',
  ])
# ---
# name: test_actions_completions[checks].2
  list([
    'all_checks',
    'check_inputs_outputs',
    'output_file_missing',
  ])
# ---
# name: test_actions_completions[environments]
  list([
    'action1',
    'action2',
    'clean_environments',
    'default_action',
  ])
# ---
# name: test_actions_completions[environments].1
  list([
    'action1',
    'action2',
    'clean_environments',
    'default_action',
  ])
# ---
# name: test_actions_completions[environments].2
  list([
    'action1',
    'action2',
    'clean_environments',
    'default_action',
  ])
# ---
# name: test_actions_completions[failing_jobs]
  list([
    'all',
    'gcc',
    'go',
    'make',
    'npm',
  ])
# ---
# name: test_actions_completions[failing_jobs].1
  list([
    'all',
    'gcc',
    'go',
    'make',
    'npm',
  ])
# ---
# name: test_actions_completions[failing_jobs].2
  list([
    'all',
    'gcc',
    'go',
    'make',
    'npm',
  ])
# ---
# name: test_actions_completions[only_python]
  list([
    'hello',
    'hi',
  ])
# ---
# name: test_actions_completions[only_python].1
  list([
    'hello',
    'hi',
  ])
# ---
# name: test_actions_completions[only_python].2
  list([
    'hello',
    'hi',
  ])
# ---
# name: test_actions_completions[parametric]
  list([
    'setup',
    'sort',
  ])
# ---
# name: test_actions_completions[parametric].1
  list([
    'setup',
    'sort',
  ])
# ---
# name: test_actions_completions[parametric].2
  list([
    'setup',
    'sort',
  ])
# ---
# name: test_actions_completions[taskrunner]
  list([
    'touch\\ a\\ file',
  ])
# ---
# name: test_actions_completions[taskrunner].1
  list([
    'touch\\ a\\ file',
  ])
# ---
# name: test_actions_completions[taskrunner].2
  list([
    'touch\\ a\\ file',
  ])
# ---
# name: test_actions_completions[trivial]
  list([
    'transform',
  ])
# ---
# name: test_actions_completions[trivial].1
  list([
    'transform',
  ])
# ---
# name: test_actions_completions[trivial].2
  list([
    'transform',
  ])
# ---
# name: test_directory_completions[--directory ]
  list([
    'examples/checks',
    'examples/environments',
    'examples/failing_jobs',
    'examples/only_python',
    'examples/parametric',
    'examples/taskrunner',
    'examples/trivial',
  ])
# ---
# name: test_directory_completions[-C ]
  list([
    'examples/checks',
    'examples/environments',
    'examples/failing_jobs',
    'examples/only_python',
    'examples/parametric',
    'examples/taskrunner',
    'examples/trivial',
  ])
# ---
